### ISSUE TITLE
Clarify import/export usage of options

### ISSUE THREAD

**Issue by shawkins** [keycloak/keycloak#29200]
### Description

Related to #28384 and other issues we need to ensure that it's not necessary to modify buildtime options to run import / export.

This may also come into play with any kc.sh functionality we add around admin account access. 

Users may also not realize that they need to specify all relevant runtime options that were supplied to their kc start.

This issue should likely cover #24191 as well.

### Motivation

It is particularly confusing for users running the import / export from the same installation as the keycloak server because there is just a sinlge optimized jar location.

### Details

For build time options we can:
- providing some way for import / export to use a different optimzed jar location
- stop allowing / using buildtime options for import / export

Related to this:
- we should ensure that the cache and all ports (the management port in particular) are disabled when running import / export

A final consideration would be to have a mode for import/export that can run against an existing server.

**Comment by shawkins** [comment]
@vmuzikar @pedroigor @mabartos @Pepo48 https://github.com/keycloak/keycloak/compare/main...shawkins:iss29200?expand=1 is a very rough draft of a proposal for running kc.sh import / export (and eventually admin account commands) against a running server.

The idea would be that relevant commands would take a new flag something like  --local (or --attach, etc.) that would disallow anything other than the immediate command options. That is export would just take export options, import would just take import, etc. I think this will provide more continuity than introducing a separate launch mechanism - also our help system should already be able to check if this flag is specified and winnow options.

So this command would have the options validated and we'd be able to used the appropriately transformed values just like normal - but it would not launch quarkus. Instead it would short-circuit to whatever form of IPC we deem acceptable for these tasks. I haven't quite filled this in the code link, but there's a placeholder.

For the IPC I'm not sure what will be desirable, so I'm roughly demoing adding a data/local-command directory to communicate via. The idea here is that the --local command launches will create directory entries like random.in (to be deleted by the server) and the server will create response entries like random.out (to be deleted by the client). The format shown here is just a simple json payload - a command field, and a map of arguments. This of course can be changed to whatever IPC interface makes sense - as there are of course more details to be worked out here, such as deleting orphaned .out files, or if we allow sensitive properties (not currently planned) then these files would need to be encrypted, etc.

The processing, shown only for export, will then create a localized ProviderFactory based upon the incoming config and then process the request. This of course will be generalized to accomodate the other commands.

The underlying assumptions are:
- we can process the arguments without launching quarkus
- the command will be working with a known set of providers. 
- the lifecycle of those localized providers should not interfer with the currently installed providers.
- import can be run on a running server, but part of the command response will indicate that a cluster restart is needed for the import to take effect

Other possibilities:
- the logic isn't defaulting to the server config, but it could

WDYT? And who else would be good to bring in on seeing if a proposal like this makes sense?

**Comment by vmuzikar** [comment]
@shawkins Thank you for creating the proposal. Using IPC for commands that require a running server seems like an interesting idea. It would provide a better UX to users removing some of the most prominent confusions, especially around re-augmentation with those commands.

What I'd avoid is to keep both approach – the current one (import/export starts a server instance), and the IPC one (--local). It might bring additional confusion.

One problem might be the Operator scenarios. We'd probably have to use some shared logic for those commands, so that the Operator is able to generate the IPC files and store them in a mounted Secret. Another alternative would be that the Operator SSH's to the pod and runs the kc.sh commands there.

One more concern might be the performance. Import and export can be quite demanding performance wise. Currently, it spins up an independent process putting the import/export stress on a single shared component – the DB. The IPC approach would put a stress on a whole KC instance/pod affecting its availability.

**Comment by shawkins** [comment]
> What I'd avoid is to keep both approach – the current one (import/export starts a server instance), and the IPC one (--local). It might bring additional confusion.

I'm mostly envisioning the usage of --local to address all of the user scenarios that are coming up with running on the same machine as the server and/or keeping track of options.

For the reason you mention below around performance we may want to keep the ability to spin up a new process as well.

Hopefully any confusion would be aliviated by good help messages :)

> One problem might be the Operator scenarios. We'd probably have to use some shared logic for those commands, so that the Operator is able to generate the IPC files and store them in a mounted Secret. Another alternative would be that the Operator SSH's to the pod and runs the kc.sh commands there.

If we stick with the ability to spin up a new process, then nothing will change for the operator.

If you want to use the local approach with the operator the possible would likely be via pod execs:
- upload the import file to the data dir
- call to kc.sh import ...
- delete the import file

And just to double check - are sockets, even with restrictions to the remote host being localhost, not allowed here?

**Comment by vmuzikar** [comment]
> And just to double check - are sockets, even with restrictions to the remote host being localhost, not allowed here?

That's a good question. In theory it increases the attack surface.

I'm thinking if it wouldn't be a more straightforward option in the end to just separate the augmentation cache for each command, or not to store it at all for import/export.

I think we need to discuss in a wider team.

**Comment by vmuzikar** [comment]
> Right, I was assuming it would be a non-starter, but wanted to double check just in case.

I think the biggest challenge with the sockets approach would be Win support. Not sure we want to have something unix specific.

> That potentially addresses the concern about build time options. It does not address the concern about runtime options nor needing to keep adding overrides to turn things off for import / export.

True, the runtime options list required for those commands is ever growing.

> Another motivation is that having a local option will remove alot of usability issues with the proposed bootstrap-admin commands and there is no concern there around performance.

Sure, the new admin bootstrap commands are a strong motivation to look into this now.

**Comment by pedroigor** [comment]
> And just to double check - are sockets, even with restrictions to the remote host being localhost, not allowed here?
> That's a good question. In theory it increases the attack surface.
> Right, I was assuming it would be a non-starter, but wanted to double check just in case.

Yeah, I agree. I also think it is not a good idea to allow issuing commands to a running server. Or are you planning to enforce the server is bound to the loopback interface when running export/import?

> That potentially addresses the concern about build time options. It does not address the concern about runtime options > nor needing to keep adding overrides to turn things off for import / export.
> True, the runtime options list required for those commands is ever growing.

I don't see the experience being so bad when running the export and import commands as we have them today. I understand the motivation and all the issues we have open/closed around this though.

One way I think we can solve this is to make sure we are more explicit in our documentation about how these commands should be run and how options should be set.

If you are setting server options (both build and runtime) using a configuration file, you won't face issues when issuing the commands as the options will be the same as when building/starting the server. Not sure if the UX is so bad here.

The real confusion, it seems, is when people start working with optimized images where you have build/runtime options used in different stages. In this case, you should keep in mind that the same options should be used when starting the server. If we are missing any runtime option to both the export/import commands (see https://github.com/keycloak/keycloak/issues/29376) we should just add it?

The fact we are now running a build during a export is something I was kinda against when we introduced it. At that time, we only had both export/import operating on runtime options and even though this was causing confusion I always considered as a necessary evil to force a specific procedure when issuing those commands.

**Comment by pedroigor** [comment]
> The draft pr does not use a socket, but rather a data directory for IPC - https://github.com/keycloak/keycloak/compare/main...shawkins:iss29200?expand=1
> 
> I was also asking if it would be acceptable to use a socket limited to loopback, but it doesn't seem like that would be an acceptable direction.

Or are you planning to enforce the server is bound to the loopback interface when running export/import? Or the HTTP listeners will still be disabled?

Do I need to stop the server manually once I'm done with import/export?

Sorry for the many questions and if I missed something in the description/previous discussions. Trying to understand how it will look like from an user PoV ...

> 
> > I don't see the experience being so bad when running the export and import commands as we have them today. I understand the motivation and all the issues we have open/closed around this though.
> 
> I didn't do an exhaustive search, but there is at least the following issues:
> 
> #24191 #20442 #28384 #13907 #10815

By quickly looking at those issues, it seems they are mostly related to using a container and perhaps related to my previous point when using optimized images and using build/runtime options in different stages.

> 
> > One way I think we can solve this is to make sure we are more explicit in our documentation about the procedure when issuing the commands and how the options should be set.
> 
> That is what we wanted for the resolution of #28384 - which I realized is actually spelled out more explicitly in this issue, but it should get addressed one way or another.
> 

And also disable caching completely when running those commands? We already disable the HTTP/management listeners, right?

> > If you are setting server options (both build and runtime) using a configuration file, you won't face issues when issuing the commands as the options will be the same as when building/starting the server. Not sure if the UX is so bad here.
> 
> There are really two issues here - I should have make that part of this issue description - see [#20442 (comment)](https://github.com/keycloak/keycloak/issues/20442#issuecomment-2101102675)
> 
> The other problem is when you try to run import / export (or the new bootstrap-admin) commands on the same vm / machine as the running server. I think we have at least one doc note saying that you shouldn't do this, but users seem to want to do it anyway...

I see. Avoiding people from running on the same VM should not prevent them to run elsewhere and potentially impact the (prod) runtime too ....

But if we are not binding to any port (HTTP, JGroups, etc), it should not be a problem, right? Perhaps only if you database has limited connections and you will either be unable to get one when importing/exporting or steal connections from the production runtime ...

> 
> > If we are missing any build/runtime option to both the export/import commands (see #29376) we should just add it?
> 
> In some ways that adds to the confusion - at least if you are trying to use the cli.

I see that but for me this is more like a documentation issue and make very explicit that any option you set to start the server should also be set when running those commands. This is specially true for provider-specific options, and those under the security, vault, feature, transaction, and database categories.

For instance, options in the cache do not make much sense ...

> 
> Leaving things basically as they are seems to look like:
> 
> * expose all? option categories for import / export

As per above, not all of them makes sense. Mainly those related to prod runtime such as caching.

> * make it explicit that users cannot run the import / export from the same vm / machine as a running server - or add more hard-coded property overrides to prevent things like the management port conflict.

I think this does not help much to prevent the problems when running an export in parallel with a production runtime. Can we disable those bindings instead? If not already.

> * and make it explicit that import / export is effectively the same thing as starting the server, but with a couple of hard-coded options so that it doesn't join the cache - so they need to use --optimize or keep all build-time options the same, and may need to supply nearly the same set of runtime options.

Sounds like a plan to me. Mainly the part about being opinionated about certain options that do not make sense at all when running those commands.

In regards to `--optimize`, perhaps we can avoid this flag by checking the server has a server image already. If so, we skip the build phase and just run the application. In case you want to force a re-build, perhaps a flag to force a build should help when running import/export ...

> 
> It will work, but it doesn't seem like a great user experience - in particular for the new bootstrap-admin commands where performance is not a concern either.

I need to keep up with the last discussions around this topic so I can't say much if the mechanics we are discussing make sense to solve both of the problems (import/export and bootstrap-admin).

**Comment by shawkins** [comment]
> Or are you planning to enforce the server is bound to the loopback interface when running export/import? Or the HTTP listeners will still be disabled?

I don't have plans currently. The draft changes presume using a directory rather than a socket.

> Do I need to stop the server manually once I'm done with import/export?

Presumably for import you will need to roll your instances yes. You would only need to stop the server for export if you don't want the server running in the first place.

> And also disable caching completely when running those commands? We already disable the HTTP/management listeners, right?
> But if we are not binding to any port (HTTP, JGroups, etc), it should not be a problem, right?

A hard-coded override for http serving port was added in past. Until KC 25 we continue to have an issue with jgroups ports (because the relevant caching options are build time in KC 24), and no we do not disable the management port.  I have been meaning to create a follow-up issue for the management port and retesting the local cache override on KC 25. 

So while we can continue to add docs and overrides - this seems like it will keep coming up over time.

Another possibly related issues are is #29010 - the root of which hasn't been determined, but it appears that import / export can't connect to the database. 

> As per above, not all of them makes sense. Mainly those related to prod runtime such as caching.

As long as there's nothing that needs changed at build time related to that category - for KC 24 and caching, this was not the case, which was the heart of the #28384 issue.

> I think this does not help much to prevent the problems when running an export in parallel with a production runtime. Can we disable those bindings instead? If not already.

Sorry if I'm not being clear - either we need to simply state that you cannot expect to run import / export / (new commands) along side a running server, or we need to do more overrides / documentation. I think we're in agreement on the latter.

> In regards to --optimize, perhaps we can avoid this flag by checking the server has a server image already. If so, we skip the build phase and just run the application. In case you want to force a re-build, perhaps a flag to force a build should help when running import/export ...

Yes this relates to one of the bullet points in the description about either not allowing or not using build-time options with import / export. Essentially this phrasing sounds like --optimized is on by default. What I envisioned in the description was to not introduce a force or any kind of build specific to import / export - if you need a different optimized image, you'd run the build command.

> I need to keep up with the last discussions around this topic so I can't say much if the mechanics we are discussing make sense to solve both of the problems (import/export and bootstrap-admin).

On #9829 we're discussing adding the ability to bootstrap admin access with commands like:

./kc.sh boostrap-admin user --username my_temp_admin

Ideally that could be run against the local server, rather than or in addition to working like import / export do currently.

It's also worth pointing out that these approaches are not mutually exclusive - for import / export we can sure up the option categories, docs, and overrides - and also offer the ability to run against a local server.

**Comment by mabartos** [comment]
@pedroigor @shawkins @vmuzikar I've briefly gone through the discussion, and I was wondering whether to use a slightly different approach.

**What are the requirements for export?** 

AFAIK it should just connect to the DB and export the KC data in some manner, with no cache involved, nothing more.
Any port should be exposed by import/export process, such as HTTP/HTTPS, mgmt interface, JGroups, or whatever.
In that case, we prevent port collisions, as these ports are not used for anything, anyway. 

I like the idea @shawkins suggested around the `--local` prop, as only immediate command options would be available.
No augmentation allowed, build time options retrieved from the persisted config source.
In that case, the props for the export would need to be substituted somehow. I'm not sure about IPC, wondering whether there's really a need for it, as it might introduce specific complexity and some 'worse' maintainability.

What about creating another persisted config source for the runtime options? As we have already the `PersistedConfigSource`(`keycloak-persisted.properties`) for the build options, we could have something similar for the runtime ones. The runtime options would be written there always when executing `start`/`start-dev` command and used only for commands with the `--local` flag or sth like that.

The config source might be removed during the shutdown. I'm aware runtime options contain sensitive information, but I'm not sure if it's a big deal when it's available only on the local machine (we could discuss).

The example for the `export` command:

1. Start the server (`./kc.sh start --db=postgres --db-username=xxx ...`) - STORE runtime options to the runtime persisted conf source
2. Run export (`./kc.sh export --dir=. --local`) - GET runtime options from the runtime persisted conf source

And users are not forced to provide the runtime options. When there's no `--local` flag, the export command would accept all the runtime options.

![export](https://github.com/keycloak/keycloak/assets/38039883/93f7c1e0-b209-434b-92ae-c34034066b15)

As the props are not changed during server processing, runtime values might be stored as they're correct during the entire server execution. It might be useful even for the admin operations, right? 

WDYT?

**Comment by vmuzikar** [comment]
Ss I'm thinking about it more, I'm more and more inclined towards keeping the current paradigm that import/export/future command (admin-bootstrap?) starts the server and just improve the UX around it. IPC seems to be adding quite a lot of complexity IMHO and if we'd keep it alongside the current approach it'd have a maintenance cost.

From my perspective, we could do at least the following improvement:
* Do not bind any ports, all interfaces are disabled (HTTP(S), management, Jgroups – possibly whole ISPN disabled).
* Always perform the reaugmentation but do not persist anything (to avoid confusion when import/export overrides some config for start/start-dev).

The main problem still remains though and that is the config options. IMHO manually selecting the ones relevant for import/export/whatever is not an ideal approach. We should look at those commands from a bigger picture – can we really take assumptions around what options are relevant for import/export/etc.? What if enabled/disabled features affect the behaviour of import/export? What if mTLS setting somehow affect it? Etc. It's no longer just about import/export but now also about admin-boostrap, and more can come in the future. 

One of the solutions might be what @mabartos proposes. It'd basically mimic the IPC approach.

> I like the idea @shawkins suggested around the `--local` prop, as only immediate command options would be available.
> No augmentation allowed, build time options retrieved from the persisted config source.

I think the `--local` was meant for the IPC but we could leverage it in this context too.  But not sure we need it at all. Would we support also some "remote" mode? I don't think so as it'd put us back to square one – how to deal with the config options.

Another alternative would be just to simply expose all build and runtime options as with start/start-dev (without `--optimized`) and cleary document that. But that's not the best UX.

**Comment by pedroigor** [comment]
@mabartos We were persisting runtime options in the past/beginning and we changed to not do that. TBH, I don't remember exactly the reasons why we moved but perhaps something around inconsistencies/complexity when mixing persisted runtime options and when they are set when issuing a command.

@vramik Is your memory better than mine and do you remember why we changed this? :)

**Comment by shawkins** [comment]
> Start the server (./kc.sh start --db=postgres --db-username=xxx ...) - STORE runtime options to the runtime persisted conf source

Storing the runtime options has some additional considerations as that could contain passwords.

> One of the solutions might be what @mabartos proposes. It'd basically mimic the IPC approach.

The IPC layer and invocation of localized providers wouldn't be a substancial amount of code - it may not be much more than this. It requires all of the same cli mechanics and it has the same issues with creating a process that needs to be local to where there is, or was, a running server - that leads to the same concern above as to why we'd still want to keep the ability to launch standalone as well.

> Another alternative would be just to simply expose all build and runtime options as with start/start-dev (without --optimized) and cleary document that. But that's not the best UX.

We can make another pass at curation, but I'm basically in favor of doing this as part of a three pronged cleanup of the current UX - https://github.com/keycloak/keycloak/issues/29200#issuecomment-2110968257 - just as a proactive measure to avoid any additional corner cases that we make more difficult by trying to hide certain options.

If there is strong opposition to IPC, then I would lean harder on the idea that we can eliminate more of our (not user's) support complexity by just not supporting running import / export / whatever on the same install as a running server requires taking down the server first - this is in fact how it is documented https://github.com/keycloak/keycloak/blame/6b6f3fcf053cd2f298de6f20499bcc543aee058e/docs/guides/server/importExport.adoc#L33 - I recalled seeing that note at one point, but assumed we weren't following that given the support on some of these issues. For docker / podman users we can point to spinning up a separate container, and of course the operator already handles this.

**Comment by vramik** [comment]
@pedroigor I'm sorry I do not have any memory about this decision.

**Comment by pedroigor** [comment]
Sorry @vramik, I was supposed to ping @vmuzikar instead. 

In any case, I think one of the key reasons to drop persisting runtime options (in addition to inconsistency/complexity when you have persisted runtime and the same options when running a command), as stated by @mabartos, is security and the fact that we don't want sensitive data stored.

**Comment by pedroigor** [comment]
> If there is strong opposition to IPC, then I would lean harder on the idea that we can eliminate more of our (not user's) support complexity by just not supporting running import / export / whatever on the same install as a running server requires taking down the server first - this is in fact how it is documented https://github.com/keycloak/keycloak/blame/6b6f3fcf053cd2f298de6f20499bcc543aee058e/docs/guides/server/importExport.adoc#L33 - I recalled seeing that note at one point, but assumed we weren't following that given the support on some of these issues. For docker / podman users we can point to spinning up a separate container, and of course the operator already handles this.

Sounds good to me. At least for me, it is not about not liking the IPC idea but if we really need more code to solve this particular problem.

The issue around running on the same server can probably be solved if we can force running using a local cache without any http/management interface.

**Comment by mposolda** [comment]
@shawkins @vmuzikar @pedroigor I did not went through whole discussion (sorry if I am repeating something you are already aware), but just FYI: 

One reason why we historically support export/import just in the offline mode (when the server is not running) is the possible inconsistency in the DB data.

For instance if the server is running, it is possible that some users are registering and/or realm administrator is changing configuration of some realm or creating new clients (or roles or any other entities). When the export/import is executed in the middle of this, there is high chance that exported or imported data could be corrupted, which can later result in the strange issues or completely broken database (EG. partially created entity in the export file etc).

Hence I would be very careful with ever supporting export/import against running server.

If we can make sure that no other DB operations are in progress when export/import is triggered (this means blocking any DB writes by admin REST API as well as various background tasks like eg `ClearExpiredEvents`), then it can work, but we should be sure to reliable blocking them.

**Comment by stianst** [comment]
-1 to allowing import/export to attach to an running server, for two reasons; that opens up a potential path for vulnerabilities and complexity, and secondly import/export does not support a consistent import/export currently if the DB is currently in use by running KC nodes

**Comment by shawkins** [comment]
@mposolda @stianst @vmuzikar the operator will perform imports while the existing nodes are doing work. This was in part to prevent the import from being a full outage event. @vmuzikar mentioned that since that import is only adding realms, it should be safe enough to keep the current approach. Are there any concerns about that?

**Comment by vmuzikar** [comment]
I would keep the RealmImport CR impl as is – import the Realm without shutting down the cluster. It only adds Realms, doesn't overwrite it so there should be almost zero risk something else is modifying the data during import. The alternative is to shutdown the whole Keycloak deployment which seems like a too expensive tradeoff. The RealmImport CR was design as a temporary workaround until we have some reasonable CRUD-enabled CRs to replace it, and we don't have any reports of issues with the current impl.

As for the import/export/future admin-bootstrap commands. Seems the way forward is to keep the current behaviour (basically starting the server with those commands). To summarize, I see three areas where we could improve this:
* Make the augmentation independent of start/start-dev (and not persisted at all) to avoid the confusing behaviour around persisting build args.
* Disable binding of any ports. Alternatively, we could add some explicit check to gracefully fail the commands (with a clear message) if server is already running.
* Improve the config options for these commands.
  * We could keep the current behaviour (having a subset of build- and runtime options for these commands) and just improve docs/help to make the behaviour more explicit.
  * Fetching the options from a running server seems no longer relevant as we assume the server is stopped in any case.
  * We could implicitly add all start commands to import/export/... to avoid making any assumptions about what's relevant for starting the server in import/export/... mode.

**Comment by shawkins** [comment]
> Make the augmentation independent of start/start-dev (and not persisted at all) to avoid the confusing behaviour around persisting build args.

@vmuzikar moving forward since we're expecting the server to be shutdown or a separate install, I don't think we need to persue augmentation changes. @pedroigor and I have a similar thought above that import / export / whatever should just always run optimized - that will help usability some what as the number of possible options for the commands will be limited. It will also be easier than trying to change how the augmentation works. The only downside is that it will break some users, but at worst they'll have to do a initial build.

**Comment by shawkins** [comment]
A more complete breakdown of the previous comment:

| Same machine as server | --optimized | Notes on running a non-server command |
| ------------- | ------------- | ------------- |
| Y  | Y  | Will use the existing build time options. Generally the server must be stopped before the non-server command is run as there can still be port conflicts. Still needs any of the relevant runtime options set, especially the DB options |
| Y  | N | Dangerous case - should probably be avoided. Same as above however any changes to build time options will trigger an auto-build that will create a different optimized image for the next server launch as well. Bootstrap operations may not require the server to be shutdown as you can override more potential conflicts, but we make that difficult by excluding the ability to set some of those via the CLI. We've concluded here that it's simpler to just state the server must not be running even for bootstrapping. |
| N | - | No port conflict concerns, but for import / export we has consistency concerns and still want all server instances to be shutdown |

**Comment by vmuzikar** [comment]
@shawkins Do you want to keep it in the 26 milestone.

**Comment by shawkins** [comment]
@vmuzikar let me take a fresh look to see what, if anything, needs done
